---
title: "Veraison Clusters"
author: "Pietro Delfino"
date: "19 dicembre 2017"
output: html_document
highlight: pygments
---

```{r setup, include=F, warning=F}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(reshape2)
library(knitr)
library(DT)
library(dplyr)
library(readxl)
library(gplots)
library(biomaRt)
library(forcats)
library(GOstats)
library(GSEABase)
library(devtools)
library(GOstatsPlus)
```

## How I created the new Annotation Families Integrated (AFI) table
I downloaded from the Biomercator output the QTLs and meta-QTLs intervals in base pairs
<br>
I used the command *bedextraxt* from *bedops* to extract the genes in the QTLs and meta-QTLs intervals
<br>
The .bed reference file of the grapevine genome used to extract the genes is the V1 annotation mapped on the 12X.2 structural annotation downloaded from https://urgi.versailles.inra.fr/Species/Vitis/Genome-Browser . I downloaded each chromosome gff file and pasted into unique file and modified to format it like .bed file
<br>
Then I executed the following commands to sort bed files and extract genes in the intervals
```{r, engine = 'bash', eval = FALSE}
for d in ./*/ ; do (cd "$d" && sed -i 's/^\<[0-9]*\>/chr&/g'  *_metaqtls ); done

for d in ./*/ ; do (cd "$d" && for f in *_metaqtls; do ~/PROGRAMS/bedops/bin/sort-bed $f > $f.sorted; done ); done

for d in ./*/ ; do (cd "$d" && split -l 1 *.sorted --additional-suffix=.ok ); done

for d in ./*/ ; do (cd "$d" && mkdir ready_for_bedextract ); done

for d in ./*/ ; do (cd "$d" && mv *.ok ready_for_bedextract ); done

for d in ./*/ ; do (cd "$d/ready_for_bedextract/" && for f in *.ok; do  mv $f `cut -f 4 $f` ; done ); done

for d in ./*/ ; do (cd "$d/ready_for_bedextract/" && for f in *; do  ~/PROGRAMS/bedops/bin/bedextract ../../../copy/V2.1_ok.bed $f > $f.bed ; done ); done
```

Then I loaded in R all the list of genes comprised in the meta-QTLs intervals

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/mqtl_and_qtl_genes_bedops/metaqtl/")
filenames <- list.files(path = ".", pattern = ".bed")
kable(filenames, align="l")
all_metaqtl <-  lapply(filenames, read.table, header = F, sep="\t", col.names = c("metaQTL","chr","start","end","gene")) ## load all the file in one command
names(all_metaqtl) <- filenames
names(all_metaqtl) <- gsub(pattern=".bed", replacement="", x=names(all_metaqtl)) ## remove .bed suffix
all_mqtl_merge <- Reduce(function(...) merge(...,by=c("gene","chr","start","end"),  all=T), all_metaqtl) ## merge the list of file in one big dataframe
colnames(all_mqtl_merge) <- c("gene","chr","start","end", names(all_metaqtl))
kable(all_mqtl_merge[1:6,1:18], align="l")
```

The QTL genes loaded were only the ones from QTLs related to phenology, identified from the pattern option in list.files command

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/mqtl_and_qtl_genes_bedops/qtl/V1_on_12x.2")
filenames <- list.files(path = ".", pattern = ("VT|F-V|VB|VE|Vr|VP|V-R|Vr-Rp|F-R"))
kable(length(filenames), align="l")
all_qtl <-  lapply(filenames, read.table, header = F, sep="\t", col.names = c("QTL","chr","start","end","gene"))
names(all_qtl) <- filenames
names(all_qtl) <- gsub(pattern=".bed", replacement="", x=names(all_qtl)) ## remove .bed suffix
all_qtl_merge <- Reduce(function(...) merge(...,by=c("gene","chr","start","end"),  all=T), all_qtl)
colnames(all_qtl_merge) <- c("gene","chr","start","end", names(all_qtl))
kable(all_qtl_merge[1:6,1:18], align="l")
```

Now I load the original AFI file as received from Sara

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
## merge qtls and metaqtls (this command I cannot do it on the laptop, too memory consuming)
# all_qtls_metaqtls_V1_on_12x.2 <- merge(all_qtl_merge, all_mqtl_merge, all=T)
## I load the file elaborated on the server
# write.table(all_qtls_metaqtls_V1_on_12x.2, file="../all_qtls_metaqtls_V1_on_12x.2.txt", row.names=F, quote=F, sep="\t", dec=".")
all_qtls_metaqtls_V1_on_12x.2 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/all_qtls_metaqtls_V1_on_12x.2.txt", header=T)
afi_original <- read_excel("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/OLD-FILES/nuovi_database/Annotation_Families_Integrated_2016.xlsx", sheet= "Annotation_Families_Integrated")
kable(head(afi_original), align="l")
## remove useless information
afi_original_reduced <- afi_original[,c(2,13)]
afi_original_reduced <- dplyr::filter(afi_original_reduced, grepl("VIT",  `Unique ID`)) ## some filtering
## merge qtls and metaqtls with afi (this commands I cannot do it on the laptop, too memory consuming)
# all_qtls_metaqtls_V1_on_12x.2_with_annotation <- merge(all_qtls_metaqtls_V1_on_12x.2, afi_original_reduced, by.x = "gene", by.y="Unique ID", all=T)
# all_qtls_metaqtls_V1_on_12x.2_with_annotation <- merge(all_qtls_metaqtls_V1_on_12x.2, afi_original_reduced, all=T)
# kable(head(all_qtls_metaqtls_V1_on_12x.2_with_annotation), align="l")

# from here on I performed different merge to attach the new transcriptomic datasets to the afi file, until i get this big file with all the necessary information
new_afi_definitive <- read_excel("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/all_qtls_metaqtls_V1_on_12x.2_with_annotation.xlsx")
dim(new_afi_definitive)
```

The table is available at https://pietrod.shinyapps.io/Annotation_Families_Integrated/ 

## Molecular Veraison

I received 6 different datasets of Pinot Noir and Cabernet Sauvignon cultivars RNA-Seq FPKM data for 3 different years of berries samples around veraison time
<br>
I load the original dataset to perform filtering and cleaning. We need to identify the *molecular veraison* for the different years and cultivars in order to align the datasets and perform clustering

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
pn.12 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/PN12_FPKM_table.txt", header=T, row.names=1)
pn.13 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/PN13_FPKM_table.tsv", header=T, row.names=1)
pn.14 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/PN14_FPKM_table.tab", header=T, row.names=1)
cs.12 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/CS12_FPKM_table.txt", header=T, row.names=1)
cs.13 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/CS13_FPKM_table.tsv", header=T, row.names=1)
cs.14 <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/CS14_FPKM_table.tab", header=T, row.names=1)
dim(pn.12)
dim(pn.13)
dim(pn.14)
dim(cs.12)
dim(cs.13)
dim(cs.14)
kable(pn.12[1:6,1:18],align="l")
```

The idea is to calculate the number of genes, in particular the *marker transitions* **early up**, that show a value of log2FC greater than 1.5 in the intervals around veraison, in order to identify what time interval shows the highest number of mt genes moving and displaying significant increase in the expression level, to be able to say when is molecular veraison occurring
<br>
I wrote an R function that starting from the raw datasets of FPKM perform filtering and cleaning according to Sara suggestions and then calculate the fold change of the mt genes between each pair of time point from T0 to T5 and return the number of genes with a log2FC higher than 1.5
<br>
The function works like this:

1. load a raw dataset and identify the total number of columns (conditions + replicates)
2. it creates a vector representing the number of columns to add to the dataset that will be filled with NA or keep according to the condition evaluated
3. the number of column to add is equal to the total number of columns + 1 and the total number of columns + total number of columns divided by 3 (if the number of replicates is 3)
4. the idea is to evaluate separately every sample and in the corresponding added column for the sample set NA if the FPKM value is less than 1 in at least 2 replicates
5. the set of new columns is then evaluated by row and if all the columns of a row show NA, the entire row is removed
6. the mean values of the remaining rows calculated by sample (3 replicates) is then obtained and a new dataframe is created with only the mean values
7. this dataframe is subsetted based on the *early up* mt genes
8. the fold change (FC) is calculated for every gene between a time point and the one before (T1/T0)
9. log2 of the FC is applied and the number of genes showing FC > 1.5 is returned

```{r, include=TRUE, warning=F, eval=F}
molecular_veraison_windows <- function(filename) {
	dat <- read.table(file = paste0("../../../nuovi_dataset/",filename), header=T, row.names=1)
	clmn.n <- length(names(dat))
	col.to.add <- (length(names(dat)) + 1) : ((length(names(dat)) + (length(names(dat))/3)))
	dat[, col.to.add] <- (ifelse(sapply(seq(1,length(names(dat)),by=3),function(i) rowSums(dat[,i:(i+2)] < 1, na.rm=T) > 1),NA,"keep"))
	dat.1 <- dat[rowSums(is.na(dat[,col.to.add]))!=ncol(dat[,col.to.add]), ]
	dat.1 <- dat.1[,1:clmn.n]
	colnames(dat.1) <- paste0(rep("t", length(names(dat.1))), rep(0:((length(names(dat.1))/3)-1),each=3))
	# number of columns per group (1-3, 4-6)
	n <- 3
	# number of groups
	n_grp <- ncol(dat.1) / n
	# column indices (one vector per group)
	idx_grp <- split(seq(dat.1), rep(seq(n_grp), each = n))
	# calculate the row means for all groups
		res <- lapply(idx_grp, function(i) {
		# subset of the data frame
		tmp <- dat.1[i]
		# calculate row means
		rowMeans(tmp, na.rm = TRUE)
	})
	# transform list into a data frame
	dat2 <- as.data.frame(res)
	# extract names of first column of each group
	names_frst <- names(dat.1)[sapply(idx_grp, "[", 1)]
	# modify column names of new data frame
	names(dat2) <- names_frst
	library(readxl)
	m.t.all <- read_excel("../../../geni_marcatori_transizioni_copy.xlsx", sheet="all")
	early_up <- m.t.all[m.t.all$marker_transitions == "early_up",]
	dat2_early_up <- dat2[rownames(dat2) %in% early_up$ID,]

	dat2_early_up <- as.data.frame(dat2_early_up)

	dat2_early_up$FC0_1 <- dat2_early_up$t1 / dat2_early_up$t0
	dat2_early_up$FC1_2 <- dat2_early_up$t2 / dat2_early_up$t1
	dat2_early_up$FC2_3 <- dat2_early_up$t3 / dat2_early_up$t2
	dat2_early_up$FC3_4 <- dat2_early_up$t4 / dat2_early_up$t3
	dat2_early_up$FC4_5 <- dat2_early_up$t5 / dat2_early_up$t4

	dat2_early_up_FC <- log2(dat2_early_up[,n_grp:(n_grp+5)])

	print(dim(dat2_early_up_FC[dat2_early_up_FC$FC0_1 > 1.5 , ]))
	print(dim(dat2_early_up_FC[dat2_early_up_FC$FC1_2 > 1.5 , ]))
	print(dim(dat2_early_up_FC[dat2_early_up_FC$FC2_3 > 1.5 , ]))
	print(dim(dat2_early_up_FC[dat2_early_up_FC$FC3_4 > 1.5 , ]))
	print(dim(dat2_early_up_FC[dat2_early_up_FC$FC4_5 > 1.5 , ]))
}
```
Example of the function

```{r, include=TRUE, warning=F}
# load the function
source("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/molecular_veraison_windows.r")
molecular_veraison_windows("PN12_FPKM_table.txt")
```

In this example (Pinot 2012) the molecular veraison is considered between T2 and T3 - 127 genes (do not consider the second number, it is just the number of columns of the dataset returned as result). The table displayed before the numbers is a snapshot of the original dataset cleaned and where the values of the replicates have been averaged
<br>
**Results for both cultivars and all years**

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
all.results <- read_excel("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/molecular_veraison-db.xlsx", sheet="all_for_R")
kable(all.results, align="l")
```

## Clusters

To perform clustering I used the software called **Clust**, available at https://github.com/BaselAbujamous/clust . It is a pyhton program which takes as input the raw datasets and perform filtering, normalization, centering and clustering. It can handle n amount of datasets with different time points or conditions
<br>
We ran **Clust** both on single datasets keeping all the time points, and on multiple datasets together after sincronizing the time points around the molecular veraison identified previously (and then reducing the number of time points)
<br>
The clustering process was run not on the entire dataset but only on the genes in the QTLs intervals, that I extracted from the new AFI file

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
pn.12.qtl.genes <- pn.12[rownames(pn.12) %in% all_qtl_merge$gene, ]
kable(pn.12.qtl.genes[1:6,1:18], align="l")
dim(pn.12.qtl.genes)
```

The number of genes in all the QTLs intervals is 8091. The same subsetting has been applied to all 6 datasets
<br>
**Clust** was then ran with the same following commands. Before on single datasets without correction around molecular veraison

```{r, engine = 'bash', eval = FALSE}
clust Data-pn_2012/ -n normalization-pn_2012.txt -r replicates_file-pn_2012.txt -fil-v 1.1 -fil-c 1 -fil-d 1 -np 18
```

And after correction when running on multiple datasets together, for example on the three years of Pinot together

```{r, engine = 'bash', eval = FALSE}
clust Data-all-pn/ -n normalization-all-pn.txt -r replicates_file-all-pn.txt -fil-v 1.1 -fil-c 1 -fil-d 3 -np 18
```

Clustering results
<br>
The following table summarizes the results obtained for the clustering before and after centering around molecular veraison

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
clust.results.before <- read_excel("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/clust_results_summary.xlsx", sheet="before")
kable(clust.results.before, align="l")
clust.results.after <- read_excel("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/clust_results_summary.xlsx", sheet="after")
kable(clust.results.after, align="l")
```

Some example plots from the clustering results of **Clust**
<br>
Before I load the genes (VIT) placed in the clusters for each separate datasets

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/Before-correction/Results_28_Nov_17/")
filenames <- list.files(path = ".",pattern=("ts.tsv"), recursive=T)
all_clust <-  lapply(filenames, read.table, head=T, sep="\t", na.strings="")
names(all_clust) <- filenames
names(all_clust) <- gsub(pattern="Results_28_Nov_17_", replacement="", x=names(all_clust))
names(all_clust) <- gsub(pattern="/Clusters_Objects.tsv", replacement="", x=names(all_clust))
## rimuovo la seconda riga (quella con scritto Genes) da tutti i df ##
all_clust <- lapply(all_clust, function(x) x[2:nrow(x),])
dim(all_clust)
```

Then I load all the data processed by **Clust** that the program used to plot the results (cleaned, filtered, centered, normalized) 
<br>
**To correct**

## Heatmaps of Pinot up-regulated genes around veraison 
```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/Before-correction/Results_28_Nov_17/")
cl.pn12 <- read.table("Results_28_Nov_17_PN-2012/Clusters_Objects.tsv", head=T, sep="\t", na.strings="")
cl.pn12 <- cl.pn12[2:nrow(cl.pn12),]
dim(cl.pn12)
cl.pn12.up <- cl.pn12[,c(3,4)]
head(cl.pn12.up)
cl.pn13 <- read.table("Results_28_Nov_17_PN-2013/Clusters_Objects.tsv", head=T, sep="\t", na.strings="")
cl.pn13 <- cl.pn13[2:nrow(cl.pn13),]
dim(cl.pn13)
cl.pn13.up <- cl.pn13[,c(6:9)]
head(cl.pn13.up)
cl.pn14 <- read.table("Results_28_Nov_17_PN-2014/Clusters_Objects.tsv", head=T, sep="\t", na.strings="")
cl.pn14 <- cl.pn14[2:nrow(cl.pn14),]
dim(cl.pn14)
cl.pn14.up <- cl.pn14[,c(8:11)]
head(cl.pn14.up)
```

Which and how many up-regulated genes are in common between all the 3 years in Pinot clusters as obtained by **Clust** on each separate datasets?

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
## I already know that these are 13 =) 
thirteen <- sort(intersect(intersect(unique(as.vector(as.matrix(cl.pn12.up[,c(1:2)]))), unique(as.vector(as.matrix(cl.pn13.up[,c(1:4)])))), unique(as.vector(as.matrix(cl.pn14.up[,c(1:4)])))))
kable(thirteen, align="l")
## Who are these genes?
kable(new_afi_definitive[new_afi_definitive$gene %in% thirteen , c(1,2,3,90:97)], align="l", row.names=F)
```

Now I retrieve the up-regulated genes as obtained by **Clust** on the 3 years datasets integrated analysis

```{r, include=TRUE, warning=F}
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/After-correction/Results_06_Dec_17/Results_06_Dec_17_pn/")
pn.up.together <- read.table("Clusters_Objects.tsv", head=T, sep="\t", na.strings="")
pn.up.together <- pn.up.together[2:nrow(pn.up.together),]
dim(pn.up.together)
head(pn.up.together)
```

Among these clusters the profiles we are interested in are the ones from clusters C3 and C4

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
## I already know that these are 62 =) 
sixtytwo <- sort(unique(as.vector(as.matrix(pn.up.together[,c(4,6)]))))
kable(sixtytwo)
## Who are these genes?
DT::datatable(new_afi_definitive[new_afi_definitive$gene %in% sixtytwo , c(1,2,3,90:97)], rownames = F, filter = "top")
```

Now I try to plot these genes as a heatmap before with raw values and then with the values normalized and centered by **Clust**
<br>
Example with Pinot 2012. I use the function I wrote to clean RNAseq datasets to clean PN12 raw dataset
<br>

**Raw FPKM**

```{r, include=TRUE, warning=F, dev="svg", fig.width=12, fig.height=9}
options(knitr.kable.NA = '')
source("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/clean_RNAseq_datasets_windows.r")
dim(pn.12)
pn.12.cleaned <- clean_RNAseq_datasets_windows("PN12_FPKM_table.txt")
dim(pn.12.cleaned)
pn.12.cleaned$ID <- rownames(pn.12.cleaned)
pn.12.cleaned.log <- log2(pn.12.cleaned[,1:10])
pn.12.cleaned.log$ID <- rownames(pn.12.cleaned.log)
## raw raw values
kable(head(pn.12.cleaned[rownames(pn.12.cleaned) %in% thirteen , -11 ]),align="l")
ggplot(melt(pn.12.cleaned[rownames(pn.12.cleaned) %in% thirteen , ], id.vars="ID"), aes(x=variable, y=value, group=ID)) + geom_line() + theme_bw() + ggtitle("pn12 thirteen raw FPKM")
heatmap.2(as.matrix(pn.12.cleaned[rownames(pn.12.cleaned) %in% thirteen , c(1:10) ]), Colv=F, trace="none",margins = c(8, 16))
kable(head(pn.12.cleaned[rownames(pn.12.cleaned) %in% sixtytwo, ]),align="l")
ggplot(melt(pn.12.cleaned[rownames(pn.12.cleaned) %in% sixtytwo, ], id.vars="ID"), aes(x=variable, y=value, group=ID)) + geom_line() + theme_bw() + ggtitle("pn12 sixtytwo raw FPKM")
heatmap.2(as.matrix(pn.12.cleaned[rownames(pn.12.cleaned) %in% sixtytwo , c(1:10) ]), Colv=F, trace="none",margins = c(8, 16), cexRow=0.4)
## log2 version
kable(head(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% thirteen , ]),align="l")
ggplot(melt(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% thirteen , ], id.vars="ID"), aes(x=variable, y=value, group=ID)) + geom_line() + theme_bw() + ggtitle("pn12 thirteen log2 FPKM")
heatmap.2(as.matrix(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% thirteen , c(1:10) ]), Colv=F, trace="none",margins = c(8, 16))
kable(head(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% sixtytwo, ]),align="l")
ggplot(melt(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% sixtytwo, ], id.vars="ID"), aes(x=variable, y=value, group=ID)) + geom_line() + theme_bw() + ggtitle("pn12 sixtytwo log2 FPKM")
is.na(pn.12.cleaned.log) <- sapply(pn.12.cleaned.log, is.infinite)
heatmap.2(as.matrix(pn.12.cleaned.log[rownames(pn.12.cleaned.log) %in% sixtytwo, c(1:10) ]), Colv=F, trace="none",margins = c(8, 16), cexRow=0.4)
```

**Data transformed by Clust**

```{r, include=TRUE, warning=F, dev="svg", fig.width=12, fig.height=9}
options(knitr.kable.NA = '')
## load data processed by Clust from single dataframe analysis
pn.12.processed.single <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/Before-correction/Results_28_Nov_17/Results_28_Nov_17_PN-2012/Processed_Data/pn_2012.txt_processed.tsv", head=T, sep="\t", na.strings="")
dim(pn.12.processed.single)
kable(head(pn.12.processed.single[pn.12.processed.single$Genes %in% thirteen, ]),align="l", row.names=F)
ggplot(melt(pn.12.processed.single[pn.12.processed.single$Genes %in% thirteen, ], id.vars="Genes"), aes(x=variable, y=value, group=Genes)) + geom_line() + theme_bw() + ggtitle("pn12 thirteen by clust single datasets")
rownames(pn.12.processed.single) <- pn.12.processed.single$Genes
heatmap.2(as.matrix(pn.12.processed.single[rownames(pn.12.processed.single) %in% thirteen, -1]), Colv=F, trace="none",margins = c(8, 16))
## load data processed by Clust from multiple dataframe analysis
pn.12.processed.together <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/After-correction/Results_06_Dec_17/Results_06_Dec_17_pn/Processed_Data/pn_2012.txt_processed.tsv", head=T, sep="\t", na.strings="")
kable(head(pn.12.processed.together[pn.12.processed.together$Genes %in% sixtytwo, ]),align="l", row.names=F)
ggplot(melt(pn.12.processed.together[pn.12.processed.together$Genes %in% sixtytwo, ], id.vars="Genes"), aes(x=variable, y=value, group=Genes)) + geom_line() + theme_bw() + ggtitle("pn12 sixtytwo by clust all 3 years")
rownames(pn.12.processed.together) <- pn.12.processed.together$Genes
heatmap.2(as.matrix(pn.12.processed.together[rownames(pn.12.processed.together) %in% sixtytwo, -1 ]), Colv=F, trace="none",margins = c(8, 16), cexRow=0.4)
```

**We decided to focus our analysis on the clusters produced by the analysis of all the 3 Pinot datasets together, that is the 62 up and 577 down regulated genes obtained by running Clust with all the 3 ds together**
<br>
Now I retrieve the down-regulated genes as obtained by **Clust** on the 3 years datasets integrated analysis

```{r, include=TRUE, warning=F}
## this file is the same as pn.up.together but I load it again anyway
setwd("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/After-correction/Results_06_Dec_17/Results_06_Dec_17_pn/")
pn.down.together <- read.table("Clusters_Objects.tsv", head=T, sep="\t", na.strings="")
pn.down.together <- pn.down.together[2:nrow(pn.down.together),]
dim(pn.down.together)
```

Among these clusters the profiles we are interested in are the ones from clusters C7 and C8

```{r, include=TRUE, warning=F}
options(knitr.kable.NA = '')
## I already know that these are 577 =) 
five77 <- sort(unique(as.vector(as.matrix(pn.down.together[,c(8,9)]))))
kable(head(five77),align="l")
## Who are these genes?
DT::datatable(new_afi_definitive[new_afi_definitive$gene %in% five77 , c(1,2,3,90:97)], rownames = F, filter = "top")
```

Plot of data processed by Clust from multiple dataframe analysis

```{r, include=TRUE, warning=F, dev="svg", fig.width=12, fig.height=9}
pn.12.processed.together <- read.table("X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/dataset_trascrittomica/nuovi_dataset/cluster_analysis-heatmap/clust/Results/After-correction/Results_06_Dec_17/Results_06_Dec_17_pn/Processed_Data/pn_2012.txt_processed.tsv", head=T, sep="\t", na.strings="")
ggplot(melt(pn.12.processed.together[pn.12.processed.together$Genes %in% five77, ], id.vars="Genes"), aes(x=variable, y=value, group=Genes)) + geom_line() + theme_bw() + ggtitle("pn12 five77 by clust all 3 years")
rownames(pn.12.processed.together) <- pn.12.processed.together$Genes
heatmap.2(as.matrix(pn.12.processed.together[rownames(pn.12.processed.together) %in% five77, -1 ]), Colv=F, trace="none",margins = c(8, 16), cexRow=0.3)
```

## Gene Ontology of genes in clusters

**Gene Ontology from BiomaRt**
<br>
Before the up-regulated (62)

```{r, include=TRUE, warning=F, dev="svg", fig.width=12, fig.height=9}
options(knitr.kable.NA = '')
plant<-useMart("plants_mart",dataset="vvinifera_eg_gene", host="plants.ensembl.org")
attributes <- listAttributes(plant)
go_attributes <- attributes[27:33,1]
kable(go_attributes, align="l")
sixtytwo.go <- getBM(attributes=c('ensembl_gene_id','chromosome_name', go_attributes),filters ='ensembl_gene_id', values = sixtytwo, mart = plant)
DT::datatable(sixtytwo.go, rownames = F, filter = "top")
ggplot(sixtytwo.go, aes(fct_infreq(goslim_goa_description))) + geom_bar() +  theme(axis.text.x=element_text(angle=90,hjust=1)) + ggtitle("GO Slim Frequency")
```

Now the gene ontologies on the down-regulated

```{r, include=TRUE, warning=F, dev="svg", fig.width=12, fig.height=9}
five77.go <- getBM(attributes=c('ensembl_gene_id','chromosome_name', go_attributes),filters ='ensembl_gene_id', values = five77, mart = plant)
DT::datatable(five77.go, rownames = F, filter = "top")
ggplot(five77.go, aes(fct_infreq(goslim_goa_description))) + geom_bar() +  theme(axis.text.x=element_text(angle=90,hjust=1)) + ggtitle("GO Slim Frequency")
```

**Custom gene ontologies enrichment from https://github.com/davfre/GOstatsPlus **

```{r, engine = 'bash', eval = FALSE}
## annotation V2 to make it look like V1 ##
wget http://genomes.cribi.unipd.it/DATA/V2/annotation/bl2go.annot_with_GO_description.txt
sed -i 's/VIT_2/VIT_/g' bl2go.annot_with_GO_description.txt
sed -i 's/\.[0-9]//g' bl2go.annot_with_GO_description.txt
```

Now the commands for preparing the file for the custom gene ontologies enrichment and the real test

```{r, include=TRUE, warning=F, dev="svg", fig.width=12}
## prepare the file from the bl2go downloaded from CRIBI
GO_v2 <- b2g_to_gsc(file = "X:/BIOINFORMATICS/Dropbox/PETER/METAQTL/QTLDB/analisi_mqtl/analisi_mqtl_070317/gene_ontology/bl2go.annot_with_GO_description.txt", organism = "vvinifera")
sixtytwo.GO.BP <- test_GO(as.character(unique(sixtytwo)) , ontology = "BP", gsc=GO_v2, pval = 0.05)
sixtytwo.GO.MF <- test_GO(as.character(unique(sixtytwo)) , ontology = "MF", gsc=GO_v2, pval = 0.05)
DT::datatable(summary(sixtytwo.GO.BP),rownames = F, filter = "top")
DT::datatable(summary(sixtytwo.GO.MF),rownames = F, filter = "top")
# go_terms <- unlist(lapply(c(sixtytwo.GO.MF),function(d){significant_terms(d, cutoff = 0.05)}))
```

Now the down-regulated genes

```{r, include=TRUE, warning=F, dev="svg", fig.width=12}
five77.GO.BP <- test_GO(as.character(unique(five77)) , ontology = "BP", gsc=GO_v2, pval = 0.05)
five77.GO.MF <- test_GO(as.character(unique(five77)) , ontology = "MF", gsc=GO_v2, pval = 0.05)
DT::datatable(summary(five77.GO.BP),rownames = F, filter = "top")
DT::datatable(summary(five77.GO.MF),rownames = F, filter = "top")
```


